main/core/src/eval/Evaluator.scala

```diff
@@ -362,15 +382,302 @@ case class Evaluator(home: os.Path,
+              //              }
+
+              val workerFut: java.util.concurrent.Future[FutureResult] = completionService.submit { () =>
+                if(failFast && someTaskFailed) {
+                  // we do not start this tasks but instead return with aborted result
+                  val newResults = group.map { task =>
+                    task -> Aborted
+                  }
+
+                  evalLog.debug(s"Skipped evaluation (because of earlier failures): ${printTerm(terminal)}")
+                  FutureResult(curWork, 0, Evaluated(newResults.toMap, Seq(), false))
+
+                } else {
+
+                  val counterMsg = nextCounterMsg()
+
+                  evalLog.debug(s"Start evaluation [${counterMsg}]: ${printTerm(terminal)}")
+                  val startTime = System.currentTimeMillis()
+
+                  val res = evaluateGroupCached(
+                    terminal,
+                    group,
+                    results,
+                    counterMsg,
+                    reporter,
+                    testReporter,
+                    logger
+                  )
+
+                  val endTime = System.currentTimeMillis()
+                  evalLog.debug(s"Finished evaluation [${counterMsg}]: ${printTerm(terminal)}")
+
+                  FutureResult(curWork, (endTime - startTime).toInt, res)
+                }
+              }
+
+              evalLog.debug(s"New future: ${workerFut} [${index + 1}/${newInProgress.size}] for task: ${printTerm(terminal)}")
+              futures = futures ++ List(workerFut -> curWork)
+          }
+        } else {
+          evalLog.debug(s"No new tasks to schedule (issuer: ${issuer})")
+        }
+      }
+
+      scheduleWork("initial request")
+
+      // Work queue management
+      // wait for finished jobs and schedule more work, if possible
+      while (futures.size > 0) {
+        evalLog.debug(s"Waiting for next future completion of ${executorService}")
+        val compFuture: Future[FutureResult] = completionService.take()
+
+        val compTask = futures.find(_._1 == compFuture).map(_._2).get
+        val compTaskName = printTerm(compTask._1)
+        evalLog.debug(s"Completed future: ${compFuture} for task ${compTaskName}")
+        futures = futures.filterNot(_._1 == compFuture)
+        try {
+          val FutureResult(
+            finishedWork,
+            time,
+            Evaluated(newResults, newEvaluated, cached)
+          ) = compFuture.get()
+
+          // Check if we failed
+          someTaskFailed = someTaskFailed || newResults.exists(task => !task._2.isInstanceOf[Success[_]])
+
+          // Update state
+          evaluated.appendAll(newEvaluated)
+          results ++= newResults
+          timings.append((finishedWork._1, time, cached))
+          inProgress = inProgress.filterNot(_ == finishedWork)
```

Comment:
Seems like we're basically using `inProgress` as a `Set` data structure: we query it's size, add to it, and remove elements from it by identity. Should we make it a `Set` instead of a `List`?

Reply:
Yes, we should. Changed.